drivers/sensor/mcp98xx/mcp98xx.c

```diff
@@ -0,0 +1,209 @@
+	uint16_t value = 0;
+	int result = mcp98xx_reg_read(dev, MCP98XX_REG_CONFIG, &value);
+	if (result == 0) {
+		value &= ~MCP98XX_REG_CONFIG_SHDN;
+		return mcp98xx_reg_write_16bit(dev, MCP98XX_REG_CONFIG, value );
+	}
+	return result;
+}
+
+static int mcp98xx_set_temperature_resolution(const struct device *dev,
+					      uint8_t resolution)
+{
+	int result;
+	uint16_t resolution_16bit = resolution;
+#if CONFIG_MCP98XX_CHIP_MCP9844
+	mcp98xx_shutdown(dev);
+#endif
+	result = mcp98xx_reg_write_16bit(dev, MCP98XX_REG_RESOLUTION, resolution_16bit);
+	return result;
+}
+
+static int mcp98xx_sample_fetch(const struct device *dev,
+				enum sensor_channel chan)
+{
+	struct mcp98xx_data *data = dev->data;
+	int result;
+#if !defined(CONFIG_MCP98XX_CONTINUOUS_CONVERSION)
+	uint16_t current_temperature;
+	uint16_t previous_temperature;
+	const struct mcp98xx_config *cfg = dev->config;
+#endif
+
+	__ASSERT_NO_MSG(chan == SENSOR_CHAN_ALL || chan == SENSOR_CHAN_AMBIENT_TEMP);
+
+#if !defined(CONFIG_MCP98XX_CONTINUOUS_CONVERSION)
+	mcp98xx_wakeup(dev);
+	int64_t starttime = k_uptime_get();
+	result = mcp98xx_reg_read(dev, MCP98XX_REG_TEMP_AMB, &current_temperature);
+	do {
+		previous_temperature = current_temperature;
+		result = mcp98xx_reg_read(dev, MCP98XX_REG_TEMP_AMB, &current_temperature);
+		k_msleep(10);
+	} while(previous_temperature == current_temperature && !mcp98xx_tconv_elapsed(cfg->resolution, starttime));
+#endif
+	result = mcp98xx_reg_read(dev, MCP98XX_REG_TEMP_AMB, &data->reg_val);
+
+#if !defined(CONFIG_MCP98XX_CONTINUOUS_CONVERSION)
+	mcp98xx_shutdown(dev);
+#endif
+	return result;
+}
+
+static int mcp98xx_channel_get(const struct device *dev,
+			       enum sensor_channel chan,
+			       struct sensor_value *val)
+	{
+	const struct mcp98xx_data *data = dev->data;
+	int temp = mcp98xx_temp_signed_from_reg(data->reg_val);
+
+	__ASSERT_NO_MSG(chan == SENSOR_CHAN_AMBIENT_TEMP);
+
+	val->val1 = temp / MCP98XX_TEMP_SCALE_CEL;
+	temp -= val->val1 * MCP98XX_TEMP_SCALE_CEL;
+	val->val2 = (temp * 1000000) / MCP98XX_TEMP_SCALE_CEL;
+
+	return 0;
+}
+
+static const struct sensor_driver_api mcp98xx_api_funcs = {
+	.sample_fetch = mcp98xx_sample_fetch,
+	.channel_get = mcp98xx_channel_get,
+#ifdef CONFIG_MCP98XX_TRIGGER
+	.attr_set = mcp98xx_attr_set,
+	.trigger_set = mcp98xx_trigger_set,
+#endif /* CONFIG_MCP98XX_TRIGGER */
+};
+
+int mcp98xx_init(const struct device *dev)
+{
+	const struct mcp98xx_config *cfg = dev->config;
+	int rc = 0;
+
+	if (!device_is_ready(cfg->i2c.bus)) {
+		LOG_ERR("Bus device is not ready");
+		return -ENODEV;
+	}
+
+	rc = mcp98xx_set_temperature_resolution(dev, cfg->resolution);
+	if (rc) {
+		LOG_ERR("Could not set the resolution of mcp98xx module");
+		return rc;
+	}
+
+#ifdef CONFIG_MCP98XX_TRIGGER
+	if (cfg->int_gpio.port) {
+		rc = mcp98xx_setup_interrupt(dev);
+	}
+#endif /* CONFIG_MCP98XX_TRIGGER */
+
+#ifdef CONFIG_MCP98XX_CONTINUOUS_CONVERSION
+	rc = mcp98xx_wakeup(dev);
```

Comment:
According to the datasheets of both the MCP9808 and MCP9844, continuous mode is the power up default. So you are now setting it to continuous when it already is. Would it be nice to add a `#else` to also shutdown the device as it is actually required according to the datasheet.

Reply:
Fixed: I put the device into shutdown after _init if not in continuous conversion mode.
Also I cleaned up/used more consistend #if !defined/#ifdef . I think I could not get around 1 #if defined().